// @authzed/authzed-node@0.13.0/dist/src/authzedapi/authzed/api/v1/experimental_service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExperimentalService = exports.BulkExportRelationshipsResponse = exports.BulkExportRelationshipsRequest = exports.BulkImportRelationshipsResponse = exports.BulkImportRelationshipsRequest = exports.BulkCheckPermissionResponseItem = exports.BulkCheckPermissionPair = exports.StreamingBulkCheckPermissionResponse = exports.BulkCheckPermissionResponse = exports.BulkCheckPermissionRequestItem = exports.BulkCheckPermissionRequest = exports.StreamingBulkCheckPermissionRequest = void 0;
// @generated by protobuf-ts 2.8.3 with parameter generate_dependencies,long_type_string,client_grpc1
// @generated from protobuf file "authzed/api/v1/experimental_service.proto" (package "authzed.api.v1", syntax proto3)
// tslint:disable
const runtime_rpc_1 = require("@protobuf-ts/runtime-rpc");
const runtime_1 = require("@protobuf-ts/runtime");
const runtime_2 = require("@protobuf-ts/runtime");
const runtime_3 = require("@protobuf-ts/runtime");
const runtime_4 = require("@protobuf-ts/runtime");
const runtime_5 = require("@protobuf-ts/runtime");
const core_1 = require("./core");
const core_2 = require("./core");
const core_3 = require("./core");
const permission_service_1 = require("./permission_service");
const status_1 = require("../../../google/rpc/status");
const core_4 = require("./core");
const struct_1 = require("../../../google/protobuf/struct");
const core_5 = require("./core");
const core_6 = require("./core");
const permission_service_2 = require("./permission_service");
// @generated message type with reflection information, may provide speed optimized methods
class StreamingBulkCheckPermissionRequest$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.StreamingBulkCheckPermissionRequest", [
            { no: 1, name: "consistency", kind: "message", T: () => permission_service_2.Consistency },
            { no: 2, name: "items", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => exports.BulkCheckPermissionRequestItem, options: { "validate.rules": { repeated: { items: { message: { required: true } } } } } }
        ]);
    }
    create(value) {
        const message = { items: [] };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.Consistency consistency */ 1:
                    message.consistency = permission_service_2.Consistency.internalBinaryRead(reader, reader.uint32(), options, message.consistency);
                    break;
                case /* repeated authzed.api.v1.BulkCheckPermissionRequestItem items */ 2:
                    message.items.push(exports.BulkCheckPermissionRequestItem.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.Consistency consistency = 1; */
        if (message.consistency)
            permission_service_2.Consistency.internalBinaryWrite(message.consistency, writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* repeated authzed.api.v1.BulkCheckPermissionRequestItem items = 2; */
        for (let i = 0; i < message.items.length; i++)
            exports.BulkCheckPermissionRequestItem.internalBinaryWrite(message.items[i], writer.tag(2, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.StreamingBulkCheckPermissionRequest
 */
exports.StreamingBulkCheckPermissionRequest = new StreamingBulkCheckPermissionRequest$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkCheckPermissionRequest$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkCheckPermissionRequest", [
            { no: 1, name: "consistency", kind: "message", T: () => permission_service_2.Consistency },
            { no: 2, name: "items", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => exports.BulkCheckPermissionRequestItem, options: { "validate.rules": { repeated: { items: { message: { required: true } } } } } }
        ]);
    }
    create(value) {
        const message = { items: [] };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.Consistency consistency */ 1:
                    message.consistency = permission_service_2.Consistency.internalBinaryRead(reader, reader.uint32(), options, message.consistency);
                    break;
                case /* repeated authzed.api.v1.BulkCheckPermissionRequestItem items */ 2:
                    message.items.push(exports.BulkCheckPermissionRequestItem.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.Consistency consistency = 1; */
        if (message.consistency)
            permission_service_2.Consistency.internalBinaryWrite(message.consistency, writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* repeated authzed.api.v1.BulkCheckPermissionRequestItem items = 2; */
        for (let i = 0; i < message.items.length; i++)
            exports.BulkCheckPermissionRequestItem.internalBinaryWrite(message.items[i], writer.tag(2, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkCheckPermissionRequest
 */
exports.BulkCheckPermissionRequest = new BulkCheckPermissionRequest$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkCheckPermissionRequestItem$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkCheckPermissionRequestItem", [
            { no: 1, name: "resource", kind: "message", T: () => core_6.ObjectReference, options: { "validate.rules": { message: { required: true } } } },
            { no: 2, name: "permission", kind: "scalar", T: 9 /*ScalarType.STRING*/, options: { "validate.rules": { string: { maxBytes: "64", pattern: "^([a-z][a-z0-9_]{1,62}[a-z0-9])?$" } } } },
            { no: 3, name: "subject", kind: "message", T: () => core_5.SubjectReference, options: { "validate.rules": { message: { required: true } } } },
            { no: 4, name: "context", kind: "message", T: () => struct_1.Struct, options: { "validate.rules": { message: { required: false } } } }
        ]);
    }
    create(value) {
        const message = { permission: "" };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.ObjectReference resource */ 1:
                    message.resource = core_6.ObjectReference.internalBinaryRead(reader, reader.uint32(), options, message.resource);
                    break;
                case /* string permission */ 2:
                    message.permission = reader.string();
                    break;
                case /* authzed.api.v1.SubjectReference subject */ 3:
                    message.subject = core_5.SubjectReference.internalBinaryRead(reader, reader.uint32(), options, message.subject);
                    break;
                case /* google.protobuf.Struct context */ 4:
                    message.context = struct_1.Struct.internalBinaryRead(reader, reader.uint32(), options, message.context);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.ObjectReference resource = 1; */
        if (message.resource)
            core_6.ObjectReference.internalBinaryWrite(message.resource, writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* string permission = 2; */
        if (message.permission !== "")
            writer.tag(2, runtime_1.WireType.LengthDelimited).string(message.permission);
        /* authzed.api.v1.SubjectReference subject = 3; */
        if (message.subject)
            core_5.SubjectReference.internalBinaryWrite(message.subject, writer.tag(3, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* google.protobuf.Struct context = 4; */
        if (message.context)
            struct_1.Struct.internalBinaryWrite(message.context, writer.tag(4, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkCheckPermissionRequestItem
 */
exports.BulkCheckPermissionRequestItem = new BulkCheckPermissionRequestItem$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkCheckPermissionResponse$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkCheckPermissionResponse", [
            { no: 1, name: "checked_at", kind: "message", T: () => core_4.ZedToken, options: { "validate.rules": { message: { required: false } } } },
            { no: 2, name: "pairs", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => exports.BulkCheckPermissionPair, options: { "validate.rules": { repeated: { items: { message: { required: true } } } } } }
        ]);
    }
    create(value) {
        const message = { pairs: [] };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.ZedToken checked_at */ 1:
                    message.checkedAt = core_4.ZedToken.internalBinaryRead(reader, reader.uint32(), options, message.checkedAt);
                    break;
                case /* repeated authzed.api.v1.BulkCheckPermissionPair pairs */ 2:
                    message.pairs.push(exports.BulkCheckPermissionPair.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.ZedToken checked_at = 1; */
        if (message.checkedAt)
            core_4.ZedToken.internalBinaryWrite(message.checkedAt, writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* repeated authzed.api.v1.BulkCheckPermissionPair pairs = 2; */
        for (let i = 0; i < message.pairs.length; i++)
            exports.BulkCheckPermissionPair.internalBinaryWrite(message.pairs[i], writer.tag(2, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkCheckPermissionResponse
 */
exports.BulkCheckPermissionResponse = new BulkCheckPermissionResponse$Type();
// @generated message type with reflection information, may provide speed optimized methods
class StreamingBulkCheckPermissionResponse$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.StreamingBulkCheckPermissionResponse", [
            { no: 1, name: "checked_at", kind: "message", T: () => core_4.ZedToken, options: { "validate.rules": { message: { required: false } } } },
            { no: 2, name: "pairs", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => exports.BulkCheckPermissionPair, options: { "validate.rules": { repeated: { items: { message: { required: true } } } } } }
        ]);
    }
    create(value) {
        const message = { pairs: [] };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.ZedToken checked_at */ 1:
                    message.checkedAt = core_4.ZedToken.internalBinaryRead(reader, reader.uint32(), options, message.checkedAt);
                    break;
                case /* repeated authzed.api.v1.BulkCheckPermissionPair pairs */ 2:
                    message.pairs.push(exports.BulkCheckPermissionPair.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.ZedToken checked_at = 1; */
        if (message.checkedAt)
            core_4.ZedToken.internalBinaryWrite(message.checkedAt, writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* repeated authzed.api.v1.BulkCheckPermissionPair pairs = 2; */
        for (let i = 0; i < message.pairs.length; i++)
            exports.BulkCheckPermissionPair.internalBinaryWrite(message.pairs[i], writer.tag(2, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.StreamingBulkCheckPermissionResponse
 */
exports.StreamingBulkCheckPermissionResponse = new StreamingBulkCheckPermissionResponse$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkCheckPermissionPair$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkCheckPermissionPair", [
            { no: 1, name: "request", kind: "message", T: () => exports.BulkCheckPermissionRequestItem },
            { no: 2, name: "item", kind: "message", oneof: "response", T: () => exports.BulkCheckPermissionResponseItem },
            { no: 3, name: "error", kind: "message", oneof: "response", T: () => status_1.Status }
        ]);
    }
    create(value) {
        const message = { response: { oneofKind: undefined } };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.BulkCheckPermissionRequestItem request */ 1:
                    message.request = exports.BulkCheckPermissionRequestItem.internalBinaryRead(reader, reader.uint32(), options, message.request);
                    break;
                case /* authzed.api.v1.BulkCheckPermissionResponseItem item */ 2:
                    message.response = {
                        oneofKind: "item",
                        item: exports.BulkCheckPermissionResponseItem.internalBinaryRead(reader, reader.uint32(), options, message.response.item)
                    };
                    break;
                case /* google.rpc.Status error */ 3:
                    message.response = {
                        oneofKind: "error",
                        error: status_1.Status.internalBinaryRead(reader, reader.uint32(), options, message.response.error)
                    };
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.BulkCheckPermissionRequestItem request = 1; */
        if (message.request)
            exports.BulkCheckPermissionRequestItem.internalBinaryWrite(message.request, writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* authzed.api.v1.BulkCheckPermissionResponseItem item = 2; */
        if (message.response.oneofKind === "item")
            exports.BulkCheckPermissionResponseItem.internalBinaryWrite(message.response.item, writer.tag(2, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* google.rpc.Status error = 3; */
        if (message.response.oneofKind === "error")
            status_1.Status.internalBinaryWrite(message.response.error, writer.tag(3, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkCheckPermissionPair
 */
exports.BulkCheckPermissionPair = new BulkCheckPermissionPair$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkCheckPermissionResponseItem$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkCheckPermissionResponseItem", [
            { no: 1, name: "permissionship", kind: "enum", T: () => ["authzed.api.v1.CheckPermissionResponse.Permissionship", permission_service_1.CheckPermissionResponse_Permissionship, "PERMISSIONSHIP_"], options: { "validate.rules": { enum: { definedOnly: true, notIn: [0] } } } },
            { no: 2, name: "partial_caveat_info", kind: "message", T: () => core_3.PartialCaveatInfo, options: { "validate.rules": { message: { required: false } } } }
        ]);
    }
    create(value) {
        const message = { permissionship: 0 };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.CheckPermissionResponse.Permissionship permissionship */ 1:
                    message.permissionship = reader.int32();
                    break;
                case /* authzed.api.v1.PartialCaveatInfo partial_caveat_info */ 2:
                    message.partialCaveatInfo = core_3.PartialCaveatInfo.internalBinaryRead(reader, reader.uint32(), options, message.partialCaveatInfo);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.CheckPermissionResponse.Permissionship permissionship = 1; */
        if (message.permissionship !== 0)
            writer.tag(1, runtime_1.WireType.Varint).int32(message.permissionship);
        /* authzed.api.v1.PartialCaveatInfo partial_caveat_info = 2; */
        if (message.partialCaveatInfo)
            core_3.PartialCaveatInfo.internalBinaryWrite(message.partialCaveatInfo, writer.tag(2, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkCheckPermissionResponseItem
 */
exports.BulkCheckPermissionResponseItem = new BulkCheckPermissionResponseItem$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkImportRelationshipsRequest$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkImportRelationshipsRequest", [
            { no: 1, name: "relationships", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => core_2.Relationship, options: { "validate.rules": { repeated: { items: { message: { required: true } } } } } }
        ]);
    }
    create(value) {
        const message = { relationships: [] };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* repeated authzed.api.v1.Relationship relationships */ 1:
                    message.relationships.push(core_2.Relationship.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* repeated authzed.api.v1.Relationship relationships = 1; */
        for (let i = 0; i < message.relationships.length; i++)
            core_2.Relationship.internalBinaryWrite(message.relationships[i], writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkImportRelationshipsRequest
 */
exports.BulkImportRelationshipsRequest = new BulkImportRelationshipsRequest$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkImportRelationshipsResponse$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkImportRelationshipsResponse", [
            { no: 1, name: "num_loaded", kind: "scalar", T: 4 /*ScalarType.UINT64*/ }
        ]);
    }
    create(value) {
        const message = { numLoaded: "0" };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* uint64 num_loaded */ 1:
                    message.numLoaded = reader.uint64().toString();
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* uint64 num_loaded = 1; */
        if (message.numLoaded !== "0")
            writer.tag(1, runtime_1.WireType.Varint).uint64(message.numLoaded);
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkImportRelationshipsResponse
 */
exports.BulkImportRelationshipsResponse = new BulkImportRelationshipsResponse$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkExportRelationshipsRequest$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkExportRelationshipsRequest", [
            { no: 1, name: "consistency", kind: "message", T: () => permission_service_2.Consistency },
            { no: 2, name: "optional_limit", kind: "scalar", T: 13 /*ScalarType.UINT32*/, options: { "validate.rules": { uint32: { lte: 10000, gte: 0 } } } },
            { no: 3, name: "optional_cursor", kind: "message", T: () => core_1.Cursor }
        ]);
    }
    create(value) {
        const message = { optionalLimit: 0 };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.Consistency consistency */ 1:
                    message.consistency = permission_service_2.Consistency.internalBinaryRead(reader, reader.uint32(), options, message.consistency);
                    break;
                case /* uint32 optional_limit */ 2:
                    message.optionalLimit = reader.uint32();
                    break;
                case /* authzed.api.v1.Cursor optional_cursor */ 3:
                    message.optionalCursor = core_1.Cursor.internalBinaryRead(reader, reader.uint32(), options, message.optionalCursor);
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.Consistency consistency = 1; */
        if (message.consistency)
            permission_service_2.Consistency.internalBinaryWrite(message.consistency, writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* uint32 optional_limit = 2; */
        if (message.optionalLimit !== 0)
            writer.tag(2, runtime_1.WireType.Varint).uint32(message.optionalLimit);
        /* authzed.api.v1.Cursor optional_cursor = 3; */
        if (message.optionalCursor)
            core_1.Cursor.internalBinaryWrite(message.optionalCursor, writer.tag(3, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkExportRelationshipsRequest
 */
exports.BulkExportRelationshipsRequest = new BulkExportRelationshipsRequest$Type();
// @generated message type with reflection information, may provide speed optimized methods
class BulkExportRelationshipsResponse$Type extends runtime_5.MessageType {
    constructor() {
        super("authzed.api.v1.BulkExportRelationshipsResponse", [
            { no: 1, name: "after_result_cursor", kind: "message", T: () => core_1.Cursor },
            { no: 2, name: "relationships", kind: "message", repeat: 1 /*RepeatType.PACKED*/, T: () => core_2.Relationship }
        ]);
    }
    create(value) {
        const message = { relationships: [] };
        globalThis.Object.defineProperty(message, runtime_4.MESSAGE_TYPE, { enumerable: false, value: this });
        if (value !== undefined)
            (0, runtime_3.reflectionMergePartial)(this, message, value);
        return message;
    }
    internalBinaryRead(reader, length, options, target) {
        let message = target ?? this.create(), end = reader.pos + length;
        while (reader.pos < end) {
            let [fieldNo, wireType] = reader.tag();
            switch (fieldNo) {
                case /* authzed.api.v1.Cursor after_result_cursor */ 1:
                    message.afterResultCursor = core_1.Cursor.internalBinaryRead(reader, reader.uint32(), options, message.afterResultCursor);
                    break;
                case /* repeated authzed.api.v1.Relationship relationships */ 2:
                    message.relationships.push(core_2.Relationship.internalBinaryRead(reader, reader.uint32(), options));
                    break;
                default:
                    let u = options.readUnknownField;
                    if (u === "throw")
                        throw new globalThis.Error(`Unknown field ${fieldNo} (wire type ${wireType}) for ${this.typeName}`);
                    let d = reader.skip(wireType);
                    if (u !== false)
                        (u === true ? runtime_2.UnknownFieldHandler.onRead : u)(this.typeName, message, fieldNo, wireType, d);
            }
        }
        return message;
    }
    internalBinaryWrite(message, writer, options) {
        /* authzed.api.v1.Cursor after_result_cursor = 1; */
        if (message.afterResultCursor)
            core_1.Cursor.internalBinaryWrite(message.afterResultCursor, writer.tag(1, runtime_1.WireType.LengthDelimited).fork(), options).join();
        /* repeated authzed.api.v1.Relationship relationships = 2; */
        for (let i = 0; i < message.relationships.length; i++)
            core_2.Relationship.internalBinaryWrite(message.relationships[i], writer.tag(2, runtime_1.WireType.LengthDelimited).fork(), options).join();
        let u = options.writeUnknownFields;
        if (u !== false)
            (u == true ? runtime_2.UnknownFieldHandler.onWrite : u)(this.typeName, message, writer);
        return writer;
    }
}
/**
 * @generated MessageType for protobuf message authzed.api.v1.BulkExportRelationshipsResponse
 */
exports.BulkExportRelationshipsResponse = new BulkExportRelationshipsResponse$Type();
/**
 * @generated ServiceType for protobuf service authzed.api.v1.ExperimentalService
 */
exports.ExperimentalService = new runtime_rpc_1.ServiceType("authzed.api.v1.ExperimentalService", [
    { name: "BulkImportRelationships", clientStreaming: true, options: { "google.api.http": { post: "/v1/experimental/relationships/bulkimport", body: "*" } }, I: exports.BulkImportRelationshipsRequest, O: exports.BulkImportRelationshipsResponse },
    { name: "BulkExportRelationships", serverStreaming: true, options: { "google.api.http": { post: "/v1/experimental/relationships/bulkexport", body: "*" } }, I: exports.BulkExportRelationshipsRequest, O: exports.BulkExportRelationshipsResponse },
    { name: "StreamingBulkCheckPermission", serverStreaming: true, options: { "google.api.http": { post: "/v1/experimental/permissions/streamingbulkcheckpermission", body: "*" } }, I: exports.StreamingBulkCheckPermissionRequest, O: exports.StreamingBulkCheckPermissionResponse },
    { name: "BulkCheckPermission", options: { "google.api.http": { post: "/v1/experimental/permissions/bulkcheckpermission", body: "*" } }, I: exports.BulkCheckPermissionRequest, O: exports.BulkCheckPermissionResponse }
]);
