// @authzed/authzed-node@0.13.0/dist/src/v1.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const grpc = __importStar(require("@grpc/grpc-js"));
const helpers_1 = require("./__utils__/helpers");
const struct_1 = require("./authzedapi/google/protobuf/struct");
const util_1 = require("./util");
const v1_1 = require("./v1");
describe('a check with an unknown namespace', () => {
    it('should raise a failed precondition', (done) => {
        const resource = v1_1.ObjectReference.create({
            objectType: 'test/somenamespace',
            objectId: 'bar',
        });
        const testUser = v1_1.ObjectReference.create({
            objectType: 'test/user',
            objectId: 'someuser',
        });
        const checkPermissionRequest = v1_1.CheckPermissionRequest.create({
            resource,
            permission: 'someperm',
            subject: v1_1.SubjectReference.create({
                object: testUser,
            }),
        });
        const client = (0, v1_1.NewClient)((0, helpers_1.generateTestToken)('v1-test-unknown'), 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED);
        client.checkPermission(checkPermissionRequest, function (err, response) {
            expect(response).toBe(undefined);
            expect(err?.code).toBe(grpc.status.FAILED_PRECONDITION);
            client.close();
            done();
        });
    });
});
describe('a check with an known namespace', () => {
    it('should succeed', (done) => {
        // Write some schema.
        const client = (0, v1_1.NewClient)((0, helpers_1.generateTestToken)('v1-namespace'), 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED, util_1.PreconnectServices.PERMISSIONS_SERVICE | util_1.PreconnectServices.SCHEMA_SERVICE);
        const request = v1_1.WriteSchemaRequest.create({
            schema: `definition test/user {}

      definition test/document {
        relation viewer: test/user
        permission view = viewer
      }
      `,
        });
        new Promise((resolve) => {
            client.writeSchema(request, function (err, response) {
                expect(err).toBe(null);
                resolve(response);
            });
        })
            .then((schemaResponse) => {
            expect(schemaResponse).toBeTruthy();
            return new Promise((resolve) => {
                // Write a relationship.
                const resource = v1_1.ObjectReference.create({
                    objectType: 'test/document',
                    objectId: 'somedocument',
                });
                const testUser = v1_1.ObjectReference.create({
                    objectType: 'test/user',
                    objectId: 'someuser',
                });
                const writeRequest = v1_1.WriteRelationshipsRequest.create({
                    updates: [
                        v1_1.RelationshipUpdate.create({
                            relationship: v1_1.Relationship.create({
                                resource: resource,
                                relation: 'viewer',
                                subject: v1_1.SubjectReference.create({
                                    object: testUser,
                                }),
                            }),
                            operation: v1_1.RelationshipUpdate_Operation.CREATE,
                        }),
                    ],
                });
                client.writeRelationships(writeRequest, function (err, response) {
                    expect(err).toBe(null);
                    resolve({ response, resource, testUser });
                });
            });
        })
            .then((vals) => {
            const { response, resource, testUser } = vals;
            expect(response).toBeTruthy();
            return new Promise((resolve) => {
                // Call check.
                const checkPermissionRequest = v1_1.CheckPermissionRequest.create({
                    resource,
                    permission: 'view',
                    subject: v1_1.SubjectReference.create({
                        object: testUser,
                    }),
                    consistency: v1_1.Consistency.create({
                        requirement: {
                            oneofKind: 'fullyConsistent',
                            fullyConsistent: true,
                        },
                    }),
                });
                client.checkPermission(checkPermissionRequest, function (err, response) {
                    expect(err).toBe(null);
                    resolve(response);
                });
            });
        })
            .then((response) => {
            const checkResponse = response;
            expect(checkResponse?.permissionship).toBe(v1_1.CheckPermissionResponse_Permissionship.HAS_PERMISSION);
            client.close();
            done();
        });
    });
    describe('with caveated relations', () => {
        it('should succeed', (done) => {
            // Write some schema.
            const client = (0, v1_1.NewClient)((0, helpers_1.generateTestToken)('v1-namespace-caveats'), 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED);
            const request = v1_1.WriteSchemaRequest.create({
                schema: `definition test/user {}

          caveat has_special_attribute(special bool) {
            special == true
          }

          definition test/document {
            relation viewer: test/user
            relation caveated_viewer: test/user with has_special_attribute
            permission view = viewer + caveated_viewer
          }
      `,
            });
            new Promise((resolve) => {
                client.writeSchema(request, function (err, response) {
                    expect(err).toBe(null);
                    resolve(response);
                });
            })
                .then((schemaResponse) => {
                expect(schemaResponse).toBeTruthy();
                return new Promise((resolve) => {
                    // Write a relationship.
                    const resource = v1_1.ObjectReference.create({
                        objectType: 'test/document',
                        objectId: 'somedocument',
                    });
                    const testUser = v1_1.ObjectReference.create({
                        objectType: 'test/user',
                        objectId: 'specialuser',
                    });
                    const writeRequest = v1_1.WriteRelationshipsRequest.create({
                        updates: [
                            v1_1.RelationshipUpdate.create({
                                relationship: v1_1.Relationship.create({
                                    resource: resource,
                                    relation: 'caveated_viewer',
                                    subject: v1_1.SubjectReference.create({
                                        object: testUser,
                                    }),
                                    optionalCaveat: v1_1.ContextualizedCaveat.create({
                                        caveatName: 'has_special_attribute',
                                    }),
                                }),
                                operation: v1_1.RelationshipUpdate_Operation.CREATE,
                            }),
                        ],
                    });
                    client.writeRelationships(writeRequest, function (err, response) {
                        expect(err).toBe(null);
                        resolve({ response, resource, testUser });
                    });
                });
            })
                .then((vals) => {
                const { response, resource, testUser } = vals;
                expect(response).toBeTruthy();
                return new Promise((resolve) => {
                    // Call check when user has special attribute.
                    const checkPermissionRequest = v1_1.CheckPermissionRequest.create({
                        resource,
                        permission: 'view',
                        subject: v1_1.SubjectReference.create({
                            object: testUser,
                        }),
                        consistency: v1_1.Consistency.create({
                            requirement: {
                                oneofKind: 'fullyConsistent',
                                fullyConsistent: true,
                            },
                        }),
                        context: struct_1.Struct.fromJson({ special: true }),
                    });
                    client.checkPermission(checkPermissionRequest, function (err, response) {
                        expect(err).toBe(null);
                        resolve(response);
                    });
                });
            })
                .then((response) => {
                const checkResponse = response;
                expect(checkResponse?.permissionship).toBe(v1_1.CheckPermissionResponse_Permissionship.HAS_PERMISSION);
                client.close();
                done();
            });
        });
    });
});
describe('Lookup APIs', () => {
    let token;
    beforeEach(async () => {
        token = (0, helpers_1.generateTestToken)('v1-lookup');
        const client = (0, v1_1.NewClient)(token, 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED);
        const request = v1_1.WriteSchemaRequest.create({
            schema: `definition test/user {}

      definition test/document {
        relation viewer: test/user
        permission view = viewer
      }
      `,
        });
        await new Promise((resolve) => {
            client.writeSchema(request, function (err, response) {
                expect(err).toBe(null);
                resolve(response);
            });
        });
        const resource = v1_1.ObjectReference.create({
            objectType: 'test/document',
            objectId: 'somedocument',
        });
        const writeRequest = v1_1.WriteRelationshipsRequest.create({
            updates: [
                v1_1.RelationshipUpdate.create({
                    relationship: v1_1.Relationship.create({
                        resource: resource,
                        relation: 'viewer',
                        subject: v1_1.SubjectReference.create({
                            object: v1_1.ObjectReference.create({
                                objectType: 'test/user',
                                objectId: 'someuser',
                            }),
                        }),
                    }),
                    operation: v1_1.RelationshipUpdate_Operation.CREATE,
                }),
                v1_1.RelationshipUpdate.create({
                    relationship: v1_1.Relationship.create({
                        resource: resource,
                        relation: 'viewer',
                        subject: v1_1.SubjectReference.create({
                            object: v1_1.ObjectReference.create({
                                objectType: 'test/user',
                                objectId: 'someuser2',
                            }),
                        }),
                    }),
                    operation: v1_1.RelationshipUpdate_Operation.CREATE,
                }),
            ],
        });
        await new Promise((resolve) => {
            client.writeRelationships(writeRequest, function (err, response) {
                expect(err).toBe(null);
                resolve(response);
            });
        });
    });
    it('can lookup subjects', (done) => {
        const client = (0, v1_1.NewClient)(token, 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED);
        const request = v1_1.LookupSubjectsRequest.create({
            resource: v1_1.ObjectReference.create({
                objectType: 'test/document',
                objectId: 'somedocument',
            }),
            permission: 'view',
            subjectObjectType: 'test/user',
            consistency: v1_1.Consistency.create({
                requirement: {
                    oneofKind: 'fullyConsistent',
                    fullyConsistent: true,
                },
            }),
        });
        const resStream = client.lookupSubjects(request);
        resStream.on('data', function (subject) {
            expect(['someuser', 'someuser2']).toContain(subject.subjectObjectId);
        });
        resStream.on('end', function () {
            client.close();
            done();
        });
        resStream.on('error', function (e) {
            client.close();
            done.fail(e);
        });
        resStream.on('status', function (status) {
            expect(status.code).toEqual(grpc.status.OK);
        });
    });
    it('can lookup resources', (done) => {
        const client = (0, v1_1.NewClient)(token, 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED);
        const request = v1_1.LookupResourcesRequest.create({
            subject: v1_1.SubjectReference.create({
                object: v1_1.ObjectReference.create({
                    objectType: 'test/user',
                    objectId: 'someuser',
                }),
            }),
            permission: 'view',
            resourceObjectType: 'test/document',
            consistency: v1_1.Consistency.create({
                requirement: {
                    oneofKind: 'fullyConsistent',
                    fullyConsistent: true,
                },
            }),
        });
        const resStream = client.lookupResources(request);
        resStream.on('data', function (response) {
            expect(response.resourceObjectId).toEqual('somedocument');
        });
        resStream.on('end', function () {
            client.close();
            done();
        });
        resStream.on('error', function (e) {
            client.close();
            done.fail(e);
        });
        resStream.on('status', function (status) {
            expect(status.code).toEqual(grpc.status.OK);
        });
    });
});
describe('a check with a negative timeout', () => {
    it('should fail immediately', (done) => {
        const resource = v1_1.ObjectReference.create({
            objectType: 'test/somenamespace',
            objectId: 'bar',
        });
        const testUser = v1_1.ObjectReference.create({
            objectType: 'test/user',
            objectId: 'someuser',
        });
        const checkPermissionRequest = v1_1.CheckPermissionRequest.create({
            resource,
            permission: 'someperm',
            subject: v1_1.SubjectReference.create({
                object: testUser,
            }),
        });
        const client = (0, v1_1.NewClient)((0, helpers_1.generateTestToken)('v1-test-unknown'), 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED, util_1.PreconnectServices.NONE, {
            interceptors: [(0, util_1.deadlineInterceptor)(-100)],
        });
        client.checkPermission(checkPermissionRequest, function (err, response) {
            expect(response).toBe(undefined);
            expect(err?.code).toBe(grpc.status.DEADLINE_EXCEEDED);
            client.close();
            done();
        });
    });
});
describe('Experimental Service', () => {
    let token;
    beforeEach(async () => {
        token = (0, helpers_1.generateTestToken)('v1-experimental-service');
        const client = (0, v1_1.NewClient)(token, 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED);
        const request = v1_1.WriteSchemaRequest.create({
            schema: `definition test/user {}

      definition test/document {
        relation viewer: test/user
        permission view = viewer
      }
      `,
        });
        await new Promise((resolve) => {
            client.writeSchema(request, function (err, response) {
                expect(err).toBe(null);
                resolve(response);
            });
        });
        client.close();
    });
    it('can bulk import relationships', async (done) => {
        const client = (0, v1_1.NewClient)(token, 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED);
        const writeStream = client.bulkImportRelationships((err, value) => {
            if (err) {
                done.fail(err);
            }
            expect(value?.numLoaded).toEqual('2');
            done();
        });
        const resource = v1_1.ObjectReference.create({
            objectType: 'test/document',
            objectId: 'somedocument',
        });
        writeStream.write(v1_1.BulkImportRelationshipsRequest.create({
            relationships: [
                v1_1.Relationship.create({
                    resource: resource,
                    relation: 'viewer',
                    subject: v1_1.SubjectReference.create({
                        object: v1_1.ObjectReference.create({
                            objectType: 'test/user',
                            objectId: 'someuser',
                        }),
                    }),
                }),
                v1_1.Relationship.create({
                    resource: resource,
                    relation: 'viewer',
                    subject: v1_1.SubjectReference.create({
                        object: v1_1.ObjectReference.create({
                            objectType: 'test/user',
                            objectId: 'someuser2',
                        }),
                    }),
                }),
            ],
        }));
        writeStream.end();
        client.close();
    });
    it('can bulk export relationships', async (done) => {
        const client = (0, v1_1.NewClient)(token, 'localhost:50051', v1_1.ClientSecurity.INSECURE_LOCALHOST_ALLOWED);
        const resource = v1_1.ObjectReference.create({
            objectType: 'test/document',
            objectId: 'somedocument',
        });
        const writeRequest = v1_1.WriteRelationshipsRequest.create({
            updates: [
                v1_1.RelationshipUpdate.create({
                    relationship: v1_1.Relationship.create({
                        resource: resource,
                        relation: 'viewer',
                        subject: v1_1.SubjectReference.create({
                            object: v1_1.ObjectReference.create({
                                objectType: 'test/user',
                                objectId: 'someuser',
                            }),
                        }),
                    }),
                    operation: v1_1.RelationshipUpdate_Operation.CREATE,
                }),
                v1_1.RelationshipUpdate.create({
                    relationship: v1_1.Relationship.create({
                        resource: resource,
                        relation: 'viewer',
                        subject: v1_1.SubjectReference.create({
                            object: v1_1.ObjectReference.create({
                                objectType: 'test/user',
                                objectId: 'someuser2',
                            }),
                        }),
                    }),
                    operation: v1_1.RelationshipUpdate_Operation.CREATE,
                }),
            ],
        });
        await new Promise((resolve) => {
            client.writeRelationships(writeRequest, function (err, response) {
                expect(err).toBe(null);
                resolve(response);
            });
        });
        const resStream = client.bulkExportRelationships(v1_1.BulkExportRelationshipsRequest.create({
            consistency: v1_1.Consistency.create({
                requirement: {
                    oneofKind: 'fullyConsistent',
                    fullyConsistent: true,
                },
            }),
        }));
        resStream.on('data', function (response) {
            expect(response.relationships).toEqual([
                {
                    relation: 'viewer',
                    resource: {
                        objectType: 'test/document',
                        objectId: 'somedocument',
                    },
                    subject: {
                        optionalRelation: '',
                        object: { objectType: 'test/user', objectId: 'someuser' },
                    },
                },
                {
                    relation: 'viewer',
                    resource: {
                        objectType: 'test/document',
                        objectId: 'somedocument',
                    },
                    subject: {
                        optionalRelation: '',
                        object: { objectType: 'test/user', objectId: 'someuser2' },
                    },
                },
            ]);
        });
        resStream.on('end', function () {
            client.close();
            done();
        });
        resStream.on('error', function (e) {
            client.close();
            done.fail(e);
        });
        resStream.on('status', function (status) {
            expect(status.code).toEqual(grpc.status.OK);
        });
    });
});
